Authorization Refresh

0xifarouk opened the pull request:
Currently if the token is expired, every request after it fails and there is no way of refreshing it since it's being done on the client init.
this PR introduces better authorization handling, with the following changes:

- token is no longer being fetched on init, because you usually create the client on the server start, you want that to be as fast as possible (you don't want to make the server start waits for a network call to finish).
- token is fetched when needed if it's not there.
- token is refreshed automatically if it's expired 10 seconds before it's actual expiry, so the client requests don't get un-authorized).
- only one token refresh happens at the same time, if multiple requests are trying to access the token while it's being refreshed, all of them wait for that single refresh and take it's result.

The refresh flow is inspired by this article:
[Building a token refresh flow with async/await and Swift Concurrency
](https://www.donnywals.com/building-a-token-refresh-flow-with-async-await-and-swift-concurrency/)

0xifarouk committed:
Authorization Refresh

fpseverino commented on Sources/AppleMapsKit/Authorization/TokenResponse.swift:
```diff
@@ -0,0 +1,24 @@
+extension TokenResponse {
+    
+    /// A boolean indicates weather to token is valid 10 seconds before it's actual expiry time.
+    var isValid: Bool {
```
I think this doesn't ever become false, as the `Date.now` is new every time and `expiresInSeconds` doesn't get updated since the response is received.

Maybe we can add a private `expirationDate` property to `TokenResponse` to which we assign `Date.now.addingTimeInterval(TimeInterval(expiresInSeconds))` when the `TokenResponse` gets initialized

0xifarouk replied:
Yes it's a very good point, sorry for that.
I will push a fix and write a test for it.

fpseverino reviewed (changes requested):
Thanks again! This was something that was on my todo list

I haven't set up CI yet, so could you execute
```
swift format --in-place --recursive --parallel .
```
to format the files?

0xifarouk commented:
@fpseverino I fixed it, and added a test against it.
But now tests takes 12 seconds to run.
I tried to use a TimeTravel concept to adjust the date using a date generator, but it made the code not understandable from TokenResponse side.

if you have some idea on how to improve it let me know.

0xifarouk commented:
I improved it, and now tests take 2 seconds.

fpseverino reviewed (approved):
Great, I'll add CI, run `swift-format` and update the DocC, then I'll release a new version.
Thanks again!

0xifarouk commented:
Thank you 🙏🏻
